README.md
# About Nodeyez

Nodeyez is a project that contains a variety of python [scripts](./scripts) to
produce images based on your Bitcoin Node

Images can be displayed to video output such as an attached screen on a
Raspberry Pi, as well as to a website dashboard for browser based access. In
addition, some scripts have support for reporting data to a local
[Blockclock Mini](https://blockclockmini.com/).

Scripts can be run on their own, or run continuously in the background as a service on system startup.

# Images

A few sample images that you can produce on your node using Nodeyez are shown below

<img src="./images/blockhashdungeon.png" width=196 /><img src="./images/lndchannelbalance.png" width=196 /><img src="./images/difficultyepoch.png" width=196 /><img src="./images/fearandgreed.png" width=196 /><img src="./images/fiatprice.png" width=196 /><img src="./images/mempoolspace.png" width=196 /><img src="./images/miner.png" width=196 /><img src="./images/satsperfiatunit.png" width=196 /><img src="./images/miningpool-luxorpool.png" width=196 /><img src="./images/sysinfo.png" width=196 /><img src="./images/whirlpoolclimix.png" width=196 /><img src="./images/whirlpoolliquidity.png" width=196 />

## Bitcoin Panels

All of these panels can work with a local Bitcoin node. At this time, information is accessed via bitcoin-cli RPC calls.

* [Art Hash](./_docs/script-arthash.md)
* [Blockhash Dungeon](./_docs/script-blockhashdungeon.md)
* [Block Height](./_docs/script-blockheight.md)
* [Block Stats](./_docs/script-blockstats.md)
* [Difficulty Epoch](./_docs/script-difficultyepoch.md)
* [Fee Estimates](./_docs/feeestimates.md)
* [Halving Countdown](./_docs/script-halving.md)
* [Inscription Mempool](./_docs/script-inscriptionmempool.md)
* [Inscription Parser](./_docs/script-inscriptionparser.md)
* [Mempool Space](./_docs/script-mempoolspace.md)
* [OP_RETURN](./_docs/script-opreturn.md)

## Lighting (LND) Panels

These panels can be configured to report on local LND based nodes, as well as remote ones over REST.

* [LND Channel Balance](./_docs/script-lndchannelbalance.md)
* [LND Channel Fees](./_docs/script-lndchannelfees.md)
* [LND Hub Account Balances](./_docs/script-lndhub.md)
* [LND Messages](./docs/script-lndmessages.md)
* [LND Ring of Fire](./_docs/script-lndringoffire.md)

## Mining Panels

Panels in support of mining pools and mining devices

* [Miner Status](./_docs/script-miner.md)
* [Braiins Pool](./_docs/script-miningpool-braiinspool.md)
* [F2 Pool](./_docs/script-miningpool-f2pool.md)
* [Luxor Pool](./_docs/script-miningpool-luxorpool.md)

## Nostr Panels

Panels specifically related to Nostr

* [Nostr.Band Stats](./_docs/script-nostrbandstats.md)

## Informational Panels

These panels only get their info from local system data

* [IP Addresses](./_docs/script-ipaddresses.md)
* [System Metrics](./_docs/script-sysinfo.md)
* [UTC Clock](./_docs/script-utcclock.md)

## Other Fun Panels

* [Compass Mining Status](./_docs/script-compassminingstatus.md)
* [Dual Image Display](./_docs/script-nodeyezdual.md)
* [Fear and Greed Index](./_docs/script-fearandgreed.md)
* [Geyser Fund Projects](./_docs/script-geyserfund.md)
* [Price of Bitcoin](./_docs/script-fiatprice.md)
* [Raretoshi](./_docs/script-raretoshi.md)
* [Sats per Fiat Unit](./_docs/script-satsperfiatunit.md)
* [Whirlpool CLI Mix Status](./_docs/script-whirlpoolclimix.md)
* [Whirlpool Liquidity](./_docs/script-whirlpoolliquidity.md)

# Installation Procedures

1. [Quick Start Installation](./_install_steps/0quickstart.md) -- Quick fast setup using install script!

2. [General Node Guidance](./_install_steps/1yournode.md)
3. [Display Screen](./_install_steps/2displayscreen.md) -- (optional)
4. [Installing Common Tools](./_install_steps/3tools.md)
5. [Nodeyez User Setup](./_install_steps/4nodeyez.md)
6. [Panel Index](./_install_steps/5panels.md)
7. [Running the Slideshow](./_install_steps/6slideshow.md) -- (optional)
8. [Website Dashboard](./_install_steps/7websitedashboard.md)
9. [Running Nodeyez Services at Startup](./_install_steps/8runatstartup.md)

10. [Nodeyez-Config Tool](./_install_steps/9nodeyezconfig.md) -- NEW! Easy configuration!

# Development Environment

1. [Architecture](./_developer/0architecture.md)
2. [Python and IDE](./_developer/1basicsetup.md)
3. [Run and Change Code](./_developer/2runandchange.md)
4. [Create New Panel](./_developer/3createpanel.md)
5. [Setting up Regtest](./_developer/4regtest.md)



CHANGELOG.md
---
title: NODEYEZ Change Log
---
# CHANGELOG

## main

**New Panels and Enhancements**

- New Panel: Geyser Fund panel spotlights a randomly selected project

**Bugfixes**

- Corrected documentation error referencing package inkspace instead of inkscape
- vicariousnetwork.getfromurlifstale now honors fileAge parameter
- vicariousbitcoin.pubkey_alias restored for LND panels
- Prevent crash of SysInfo panel when low CPU load and low resolution output

## 23.08

Released: 2023-08-29

Tagged Commit: [781f1ee18b919548eebfebaa74a6a5259e80635e](https://github.com/vicariousdrama/nodeyez/commit/781f1ee18b919548eebfebaa74a6a5259e80635e)

**New Panels and Enhancements**

- Difficulty Epoch panel enhanced for pruned node support to estimate times
- OP_RETURN panel now has more exclusion patterns by default
- New Panel: Fee Estimates for graphing Bitcoin Fee Estimates
- New Panel: LND Messages to display those received in invoices and keysends

**General**

- New utility script scaffolding new panel using templates
- New helper functions for top and bottom anchored text
- New helper functions for wrapping text to a width and calculating multiline text height

**Bugfixes**

- Fixed logic in removeOldImages. Prior to this release it would remove all files of the given type (default png) from the directory from which the script was being run.


## 23.07

Released: 2023-07-30

Tagged Commit: [07bab8ea03371f3baadfbb8b52ddba6c5eaa4b5a](https://github.com/vicariousdrama/nodeyez/commit/07bab8ea03371f3baadfbb8b52ddba6c5eaa4b5a)

**New Panels and Ehancements**

- Blockstats panels for feerates now reports the low value in addition to high and avg
- Difficulty Epoch panel now shows key block height labels and dithered outline for unexpected non-mined
- Nostr.Band Stats panel graphs the Active Pubkeys, Badge Definitions and Zaps per Day seen over the previous 2 months

**General**

- Removed Twitter references
- Added maze.html serverless sample (https://nodeyez.com/maze.html)
- Reverted Python dependency to support 3.7
- Installation now gives guidance on setting up UFW and fail2ban instead of assuming it was present. UFW is called upon to enable the website dashboard port for local network access
- Mitigating partial RaspiBolt setup (Bitcoin installed, but not NGINX)
- Updates to resolve deprecated and removed functions in Pillow

**Bugfixes**

- Fixed Bitcoin CLI handling of getblockhash 
- Fixed halving if block retrieval failed
- Fixed install script failure
- Fixed some documentation images and links

## 23.06

Released: 2023-06-28

Tagged Commit: [bd4fb43d249ec1f69414467f10fe13c0dca826cc](https://github.com/vicariousdrama/nodeyez/commit/bd4fb43d249ec1f69414467f10fe13c0dca826cc)

**New Panels and Ehancements**

- Blockstats panel now supports sub-panels for input and output script types
- Compass Hardware moved to deprecated-scripts and service file deleted
- Halving panel scaled better and informs number of blocks to go, estimated times
- Inscription Mempool now derives more file extensions and displays up to 24 images
- LNDHub panel now supports pagination and will show up to 6 accounts per page by default
- Renamed rofstatus to lndringoffire
- Renamed luxor-mining-hashrate to luxorpool
- Renamed mempoolblocks to mempoolspace
- Renamed ordinals to inscriptionparser
- Renamed satsperusd to satsperfiatunit
- Renamed ipaddress to ipaddresses to prevent a common package import problem
- Reorganized luxor dependencies
- Merged MinerBraiins and MinerMicroBT panels into new generalized Miner panel
- OP Return panel can be configured to exclude common phrases to improve signal to noise ratio
- Calls to get price info now saved locally and only update once an hour.
- Sats per USD can be rendered as triangle or s symbol in addition to square or circle
- Whirlpool Liquidity panel should indicate number of inputs required (default 5)

**General**

- New NodeyezPanel class can be used for bootstrapping new panels
- Refactored the following panels to use the new class format
  - arthash
  - blockhashdungeon
  - blockheight
  - blockstats
  - compassminingstatus
  - difficultyepoch
  - fearandgreed
  - fiatprice
  - gasprice
  - halving
  - inscriptionmempool
  - inscriptionparser (fka ordinals)
  - ipaddresses (fka ipaddress)
  - lndchannelbalance
  - lndchannelfees
  - lndhub
  - lndringoffire (fka rofstatus)
  - mempoolspace (fka mempoolblocks, mempool.json config)
  - miner (from minerbraiins and minermicrobt)
  - miningpool-braiinspool (fka braiinspool, slushpool)
  - miningpool-f2pool (fka f2pool)
  - miningpool-luxorpool (fka luxor-mining-hashrate w/o service and luxor.json config)
  - nodeyezdual
  - opreturn
  - raretoshi
  - satsperfiatunit (fka satsperusd)
  - sysinfo
  - utcclock
  - whirlpoolclimix
  - whirlpoolliquidity
- Improved support for nested fields in charting via dot notation
- Bitcoin related services no longer fail startup if there is no Bitcoin
- Bitcoin panels can now get data from remote bitcoin servers via REST (JSON-RPC)
- System Info icons have been encapsulated in VicariousIcons class
- Sample config files updated with changed and new property names with references to former names

**Bugfixes**

- Blockhash Dungeon will no longer show duplicate logos in the maze
- Blockstats label for Segwit Prevalence corrected
- Braiinspool fix divide by zero bug
- Fiat Price fix to render with price down color if price has declined since last check
- Ordinals fix logging for active blocklist when request block higher than vetted
- Halving fix off by 1 calculation of new halving period
- Inscription Mempool fix to support multiple inscriptions in one transaction
- LNDHub fix crashing when checking payment by user

## 23.04

Released: 2023-04-13

Tagged Commit: [3b6208a46084c1b1cf2f21b5e01febf33fed2bd7](https://github.com/vicariousdrama/nodeyez/commit/3b6208a46084c1b1cf2f21b5e01febf33fed2bd7)

**New Panels and Enhancements**

- Renamed arthashdungeon to blockhashdungeon

scripts/f2pool.py
#! /usr/bin/env python3
from datetime import datetime
from PIL import Image, ImageDraw, ImageColor
from os.path import exists
import json
import math
import subprocess
import sys
import time
import vicariousbitcoin
import vicarioustext
import vicariousnetwork

def gethashratestring(ihashrate):
    hashrate = ihashrate
    hashdesc = "h/s"
    while (hashrate > 1000.0) and (hashdesc != "Ph/s"):
        hashrate = hashrate/1000.0
        if hashdesc == "Ph/s":
            hashdesc = "Eh/s"
        if hashdesc == "Th/s":
            hashdesc = "Ph/s"
        if hashdesc == "Gh/s":
            hashdesc = "Th/s"
        if hashdesc == "Mh/s":
            hashdesc = "Gh/s"
        if hashdesc == "Kh/s":
            hashdesc = "Mh/s"
        if hashdesc == "h/s":
            hashdesc = "Kh/s"
    hashfmt = str(format(hashrate, ".2f")) + " " + hashdesc
    return hashfmt

def getaccounthashrate(accountinfo):
    hashrate = accountinfo["hashrate"]
    return gethashratestring(hashrate)

def gethighesthashrate(accountinfo):
    highesthash = accountinfo["hashrate"]
    for key in accountinfo["hashrate_history"]:
        value = accountinfo["hashrate_history"][key]
        currentvalue = float(value)
        if currentvalue > highesthash:
            highesthash = currentvalue
    return highesthash

def getlowesthashrate(accountinfo):
    lowesthash = accountinfo["hashrate"]
    for key in accountinfo["hashrate_history"]:
        value = accountinfo["hashrate_history"][key]
        currentvalue = float(value)
        if currentvalue < lowesthash:
            lowesthash = currentvalue
    return lowesthash

def createimage(accountinfo, width=480, height=320):
    headerheight = 30
    footerheight = 15
    hashheight = (height - headerheight - footerheight) * .4
    rewardheight = (height - headerheight - footerheight) * .5
    im = Image.new(mode="RGB", size=(width, height), color=colorBackground)
    draw = ImageDraw.Draw(im)
    # Header
    vicarioustext.drawcenteredtext(draw, "F2 Pool Summary", 24, int(width/2), int(headerheight/2), colorTextFG, True)
    # Hashrate
    hashrate = getaccounthashrate(accountinfo)
    vicarioustext.drawcenteredtext(draw, "Hashrate", 16, (width/4*1), (headerheight + (hashheight/2) - 24), colorTextFG)
    vicarioustext.drawcenteredtext(draw, hashrate, 24, (width/4*1), (headerheight + (hashheight/2)), colorDataValue)
    # Yesterday and Today value
    earningspad = 24
    value_last_day = str(int(float(accountinfo["value_last_day"]) * 100000000)) + " sats"
    value_today = str(int(float(accountinfo["value_today"]) * 100000000)) + " sats"
    vicarioustext.drawcenteredtext(draw, "Earnings Yesterday", 16, (width/4*3), (headerheight + (hashheight/2) - 24 - earningspad), colorTextFG)
    vicarioustext.drawcenteredtext(draw, value_last_day, 24, (width/4*3), (headerheight + (hashheight/2) - earningspad), colorDataValue)
    vicarioustext.drawcenteredtext(draw, "Earnings Today", 16, (width/4*3), (headerheight + (hashheight/2) - 24 + earningspad), colorTextFG)
    vicarioustext.drawcenteredtext(draw, value_today, 24, (width/4*3), (headerheight + (hashheight/2) + earningspad), colorDataValue)
    # Hashrate History
    highesthashrate = gethighesthashrate(accountinfo)
    lowesthashrate = getlowesthashrate(accountinfo)
    labelwidth = math.floor(width / 5)
    graphedge = 3
    vicarioustext.drawcenteredtext(draw, "24 Hour Hashrate", 16, int(width/2), (headerheight+hashheight), colorTextFG)
    charttop = headerheight + hashheight + 12
    chartleft = labelwidth + graphedge
    chartright = width - graphedge
    chartbottom = height - footerheight - graphedge
    # - chart border
    draw.line(xy=[chartleft, charttop, chartleft, chartbottom],fill=colorGraphLineLight,width=1)
    draw.line(xy=[chartleft, chartbottom, chartright, chartbottom],fill=colorGraphLineLight,width=1)
    draw.line(xy=[chartleft, charttop, chartright, charttop],fill=colorGraphLineDark,width=1)
    draw.line(xy=[chartright, charttop, chartright, chartbottom],fill=colorGraphLineDark,width=1)
    # - dashed line background
    chart0  = int(math.floor(charttop))
    chart25 = int(math.floor(charttop + ((chartbottom - charttop)/4*1)))
    chart50 = int(math.floor(charttop + ((chartbottom - charttop)/4*2)))
    chart75 = int(math.floor(charttop + ((chartbottom - charttop)/4*3)))
    chart100  = int(math.floor(charttop + ((chartbottom - charttop))))
    for i in range(chartleft, chartright, 10):
        draw.line(xy=[i,chart25,i+1,chart25],fill=colorGraphLineDark,width=1)
        draw.line(xy=[i,chart50,i+1,chart50],fill=colorGraphLineDark,width=1)
        draw.line(xy=[i,chart75,i+1,chart75],fill=colorGraphLineDark,width=1)
    # - left labels
    hashrate25 = lowesthashrate + ((highesthashrate - lowesthashrate)/4*3)
    hashrate50 = lowesthashrate + ((highesthashrate - lowesthashrate)/4*2)
    hashrate75 = lowesthashrate + ((highesthashrate - lowesthashrate)/4*1)
    vicarioustext.drawrighttext(draw, gethashratestring(highesthashrate), 12, labelwidth, chart0, colorTextFG)
    vicarioustext.drawrighttext(draw, gethashratestring(hashrate25), 12, labelwidth, chart25, colorTextFG)
    vicarioustext.drawrighttext(draw, gethashratestring(hashrate50), 12, labelwidth, chart50, colorTextFG)
    vicarioustext.drawrighttext(draw, gethashratestring(hashrate75), 12, labelwidth, chart75, colorTextFG)
    vicarioustext.drawrighttext(draw, gethashratestring(lowesthashrate), 12, labelwidth, chart100, colorTextFG)
    # - data plot
    entrynum = 0
    plotbuf=2
    entrycount = len(accountinfo["hashrate_history"].keys())
    if entrycount > 0:
        datawidth = ((chartright - chartleft) / entrycount)
        datawidthi = int(math.floor(datawidth))
        for key in accountinfo["hashrate_history"]:
            entrynum = entrynum + 1
            datax = chartleft + int(math.floor(entrynum * datawidth))
            datapct = 0
            value = accountinfo["hashrate_history"][key]
            if highesthashrate > lowesthashrate:
                datapct = (value - lowesthashrate)/(highesthashrate - lowesthashrate)
            plottop = chartbottom - int(math.floor((chartbottom-charttop)*datapct))
            colorDotFill = colorHashDotFill
            colorDotOutline = colorHashDotOutline
            if value < hashrateLowThreshold:
                colorDotFill = colorHashDotFillLow
                colorDotOutline = colorHashDotOutlineLow
            if value == 0:
                colorDotFill = colorHashDotFillZero
                colorDotOutline = colorHashDotOutlineZero
            draw.ellipse(xy=[(datax-plotbuf,plottop-plotbuf),(datax+datawidthi+plotbuf,plottop+datawidthi+plotbuf)],fill=colorDotFill,outline=colorDotOutline,width=1)
        # moving average line
        entrynum = 0
        ma = [-1,-1,-1]
        olddatax = -1
        olddatay = -1
        for key in accountinfo["hashrate_history"]:
            entrynum = entrynum + 1
            value = accountinfo["hashrate_history"][key]
            if ma[0] == -1:
                ma[0] = value
            elif ma[1] == -1:
                ma[1] = value
            else:
                ma[2] = value
                mavg = (ma[0] + ma[1] + ma[2]) / 3
                datax = chartleft + int(math.floor(entrynum * datawidth))
                datapct = 0
                if highesthashrate > lowesthashrate:
                    datapct = (mavg - lowesthashrate)/(highesthashrate - lowesthashrate)
                datay = chartbottom - int(math.floor((chartbottom-charttop)*datapct))
                if olddatax != -1:
                    draw.line(xy=[(olddatax,olddatay),(datax,datay)],fill=colorMovingAverage,width=3)
                olddatax = datax
                olddatay = datay
                ma[0] = -1 #ma[1]
                ma[1] = -1 #ma[2]
                ma[2] = -1


    # Date and Time
    dt = "as of " + vicarioustext.getdateandtime()
    vicarioustext.drawbottomrighttext(draw, dt, 12, width, height, colorTextFG)
    # Save to file
    im.save(outputFile)
    im.close()


if __name__ == '__main__':
    # Defaults
    configFile="../config/f2pool.json"
    outputFile="../imageoutput/f2pool.png"
    account = "--put-your-account-name-in---nodeyez/config/f2pool.json"
    useTor = True
    hashrateLowThreshold = 60000000000000 # 60 TH is 60,000,000,000,000 or 60 followed by 12 zeros
    width=480
    height=320
    sleepInterval=600
    colorDataValue=ImageColor.getrgb("#4040ff")
    colorHashDotFill=ImageColor.getrgb("#4040ff")
    colorHashDotFillZero=ImageColor.getrgb("#ff4040")
    colorHashDotFillLow=ImageColor.getrgb("#ffff40")
    colorHashDotOutline=ImageColor.getrgb("#0000ff")
    colorHashDotOutlineZero=ImageColor.getrgb("#ff0000")
    colorHashDotOutlineLow=ImageColor.getrgb("#ffff00")
    colorMovingAverage=ImageColor.getrgb("#40ff40")
    colorGraphLineLight=ImageColor.getrgb("#a0a0a0")
    colorGraphLineDark=ImageColor.getrgb("#606060")
    colorTextFG=ImageColor.getrgb("#ffffff")
    colorBackground=ImageColor.getrgb("#000000")
    # Require config
    if not exists(configFile):
        print(f"You need to make a config file at {configFile} to set your account information")
        exit(1)
    # Override config
    if exists(configFile):
        with open(configFile) as f:
            config = json.load(f)
        if "f2pool" in config:
            config = config["f2pool"]
        if "account" in config:
            account = config["account"]
        if "useTor" in config:
            useTor = config["useTor"]
        if "outputFile" in config:
            outputFile = config["outputFile"]
        if "hashrateLowThreshold" in config:
            hashrateLowThreshold = config["hashrateLowThreshold"]
        if "width" in config:
            width = int(config["width"])
        if "height" in config:
            height = int(config["height"])
        if "sleepInterval" in config:
            sleepInterval = int(config["sleepInterval"])
            sleepInterval = 300 if sleepInterval < 300 else sleepInterval # minimum 5 minutes, access others
        if "colorDataValue" in config:
            colorDataValue = ImageColor.getrgb(config["colorDataValue"])
        if "colorHashDotFill" in config:
            colorHashDotFill = ImageColor.getrgb(config["colorHashDotFill"])
        if "colorHashDotFillZero" in config:
            colorHashDotFillZero = ImageColor.getrgb(config["colorHashDotFillZero"])
        if "colorHashDotFillLow" in config:
            colorHashDotFillLow = ImageColor.getrgb(config["colorHashDotFillLow"])
        if "colorHashDotOutline" in config:
            colorHashDotOutline = ImageColor.getrgb(config["colorHashDotOutline"])
        if "colorHashDotOutlineZero" in config:
            colorHashDotOutlineZero = ImageColor.getrgb(config["colorHashDotOutlineZero"])
        if "colorHashDotOutlineLow" in config:
            colorHashDotOutlineLow = ImageColor.getrgb(config["colorHashDotOutlineLow"])
        if "colorMovingAverage" in config:
            colorMovingAverage = ImageColor.getrgb(config["colorMovingAverage"])
        if "colorGraphLineLight" in config:
            colorGraphLineLight = ImageColor.getrgb(config["colorGraphLineLight"])
        if "colorGraphLineDark" in config:
            colorGraphLineDark = ImageColor.getrgb(config["colorGraphLineDark"])
        if "colorTextFG" in config:
            colorTextFG = ImageColor.getrgb(config["colorTextFG"])
        if "colorBackground" in config:
            colorBackground = ImageColor.getrgb(config["colorBackground"])
    # Check for single run
    if len(sys.argv) > 1:
        if sys.argv[1] in ['-h','--help']:
            print(f"Prepares a summary of F2 Pool hashing and earnings for 24 hours")
            print(f"Usage:")
            print(f"1) Call without arguments to run continuously using the configuration or defaults")
            print(f"2) Passing any arguments other then -h or --help will run once and exit")
            print(f"You may specify a custom configuration file at {configFile}")
        else:
            accountinfo = vicariousnetwork.getf2poolaccountinfo(useTor, account)
            createimage(accountinfo,width,height)
        exit(0)
    # Loop
    while True:
        accountinfo = vicariousnetwork.getf2poolaccountinfo(useTor, account)
        createimage(accountinfo,width,height)
        print(f"sleeping for {sleepInterval} seconds")
        time.sleep(sleepInterval)
